### DISCUSSION TITLE
Truly hide windows—impossible or not?

### DISCUSSION THREAD

**Discussion by salvalcantara** [nikitabobko/AeroSpace#1008] [general]
As noted in the [docs](https://nikitabobko.github.io/AeroSpace/guide.html#emulation-of-virtual-workspaces):

> For better or worse, macOS doesn’t allow to place windows outside the visible area entirely. You will still be able to see a few pixels of "hidden" windows in the bottom right corner of your screen. That means, that if AeroSpace crashes badly you will still be able to manually "unhide" the windows by dragging these few pixels to the center of the screen.

For me that is not a big deal but this gets _especially noticeable_—probably an exaggeration 😆—when enabling gaps.

It was indeed only after I did so that I realized the _hidden_ windows showing up in the **bottom right corner** of my screen. This is mostly _cosmetics_ I know (and from the above statement I get that _it might be actually impossible to avoid_), but raising the issue anyway just in case there is some hope / unexplored alternative that can do the trick in a better way. E.g.,

- Allow users to minimize the number of pixels shown for the hidden windows. With just one pixel they would really go unnoticed while still probably enough for _unhiding_ them in case of a crash. 😎  Otherwise, do it _at your own risk_ as they say. If some people (not we) are fine with disabling SIP for using a window manager I'm pretty sure many (if not all) will be willing to take the risk... 😉  

And, to show something, here you are how my desktop looks like for empty/non-empty workspaces, respectively:

<img width="2560" alt="Screenshot 2023-12-19 at 07 13 42" src="https://github.com/nikitabobko/AeroSpace/assets/6698989/e134573a-b928-4948-930d-b3aa03247c93">
<img width="2560" alt="Screenshot 2023-12-19 at 07 23 13" src="https://github.com/nikitabobko/AeroSpace/assets/6698989/4c043e00-4486-4951-91c5-e212b6e6bbb5">

**Comment by januz** [comment]
I was wondering the same. I would be fine with the proposed solution here, i.e., reducing the number of pixels to 1 (or having to restart if `AeroSpace` or using other tools to move the windows back to the visible space). Alternatively, could `AeroSpace` cover the other workspaces with a wallpaper?

**Comment by herself** [comment]
Also interested in this! Really like this one, and the window part lurking in the right corner is the only thing that's annoying me 😄 

Looks like this might be hackable by playing around [in the code](https://github.com/nikitabobko/AeroSpace/blob/0ce757b970a3cfd95a420eeba0bfe3b38a2f3609/src/tree/MacWindow.swift#L101), I'll try to give a spin later on...

**Comment by vrde** [comment]
👋 i3 veteran, AeroSpace (and MacOs) newbie here. First of all, I'm very impressed by this tool! When I moved to MacOs I lost hope to find anything similar to i3. I installed it yesterday and I already feel home :)

I also agree that issue is more a nuisance than an actual problem (actually there is only **one** application that is giving me a small issue, that is Photoshop, it flickers on the bottom right corner, I'm wondering if I can disable Photoshop to manage its own window).

MacOs has a native "hide" function (I guess we all know `⌘+H`), would this work? Or using an extra desktop as a "closet" where to move unused windows?

**Comment by nikitabobko** [comment]
@vrde please report a separate issue

It'd be handy to attach:
**1.** Screen recording. Does the window jump off the right corner?
**2.** What happens if you disable/quit AeroSpace and move it to the bottom right corner manually? Does jump off in the same way?
**3.** I'm not a photoshop user, so: is it possible to have several Photoshop windows or Photoshop always shows only a single window? (it's needed to understand whether it's possible to use "hide application" instead of moving window to the bottom right corner)

> I'm wondering if I can disable Photoshop to manage its own window

**4.** I think the relevant issue is https://github.com/nikitabobko/AeroSpace/issues/2

**Comment by rafo** [comment]
A workaround could be to use the macOS native and rather unknown ```cmd+alt+h``` shortcut, which hides all windows **but** the active window. Now you can ```cmd+tab``` to those windows you want to add. All other windows remain hidden without annoying pixels.

**Comment by juansalvatore** [comment]
I'm also interested in this! I think this tool is awesome and with this edge case handled in a more elegant way it would be perfect. This makes the window switch a bit annoying as something weird always happens in the bottom down corner, and it has conditioned me to remove all viewport paddings to hide the window corners, which on bigger monitors is a bit more of an issue as I'd like to have narrower windows. Hopefully there's a way to bypass that macos restriction or make them less visible.

**Comment by nikitabobko** [comment]
# Update

The issue is partially fixed in 0.14.0. Hidden windows in the corners are now less noticeable, now they take only 1-pixel vertical line in the bottom right or left corner.

Before:
<img width="421" alt="Screenshot 2024-08-03 at 22 57 01" src="https://github.com/user-attachments/assets/5b411437-118c-415c-9d57-4caf78b8687d">
After:
<img width="354" alt="Screenshot 2024-08-03 at 22 56 51" src="https://github.com/user-attachments/assets/976e00bb-2efb-4285-b803-782fc1be11ed">

**Comment by tobiasgiese** [comment]
Do we have to configure sth here? My hidden windows are still visible

This is a screenshot of my bottom right corner:

<img width="788" alt="image" src="https://github.com/user-attachments/assets/d281a68e-7d2a-415f-96b6-2d4737917bb4">

An this is my display arrangement:

<img width="417" alt="image" src="https://github.com/user-attachments/assets/a3c8441d-9361-4a4a-acdc-58245304670b">

On my bottom left corner (i.e., my macbook to the left of my monitor) it is working as expected 🤔

**Comment by salvalcantara** [comment]
@nikitabobko Works like a charm for me, huge thanks!

**Comment by tobiasgiese** [comment]
@nikitabobko I think I know why I can still see the windows: I have the Dock on the right side.

<img width="142" alt="image" src="https://github.com/user-attachments/assets/ad85987a-4172-40ff-bb8c-7c1b5d35b69c">

Enabling auto-hiding the dock removes the windows

<img width="86" alt="image" src="https://github.com/user-attachments/assets/02f4b242-d1fe-4c4e-a921-7e2ecf301e8f">

**Comment by TimAnthonyAlexander** [comment]
Hi there, I seem to have a very related bug, where an application that is bound to workspace B, and then workspace B is bound to monitor 2, is shown (when inactive) with full width on monitor 3.
How to fix this? It is very visible, only just happened.
![Screenshot 2024-08-13 at 15 45 11](https://github.com/user-attachments/assets/8cb3a310-7166-43a9-8df2-7e0b531578ba)

Monitor 3, showing an application from workspace b (bound to monitor 2), covering the full bottom of the monitor.

I have restarted my Mac, the window manager, Aerospace, changing settings, nothing seemed to help.

**Comment by tobiasgiese** [comment]
@TimAnthonyAlexander how is your monitor arrangement? Have you followed [the proper monitor arrangement docs](https://nikitabobko.github.io/AeroSpace/guide#proper-monitor-arrangement)?

**Comment by TimAnthonyAlexander** [comment]
Hi there, I'm sorry, now I understood that part of the documentation. Honestly I didn't understand that particular part when reading the docs the first time. Problem is solved. Thanks!

**Comment by marcosmoura** [comment]
One workaround here would be rendering a fake window acting as a wallpaper, above the "hidden" windows. I don't know how good this idea would be for the Aerospace core code, but I managed to create something using Hammerspoon:
Leaving it here in case it helps anyone:
[wallpaper.lua](https://github.com/marcosmoura/dotfiles/blob/main/dotfiles/.config/hammerspoon/config/window-manager/aerospace/wallpaper.lua#L139)

**Comment by bruceblore** [comment]
I personally don't really notice the tiny lines in the corners, but the fact that these windows are still getting rendered because they're still "visible" is causing high CPU usage from WindowServer.

**Comment by exsesx** [comment]
I still don't get why we can't just hide windows, for me that would be the game changer, for now I'm using https://github.com/wojciech-kulik/FlashSpace
